src/dfx/src/commands/mod.rs

```diff
@@ -71,43 +73,249 @@ pub enum Command {
-        Command::Quickstart(v) => quickstart::exec(env, v),
-        Command::Remote(v) => remote::exec(env, v),
-        Command::Replica(v) => replica::exec(env, v),
-        Command::Schema(v) => schema::exec(v),
-        Command::Sns(v) => sns::exec(env, v),
-        Command::Start(v) => start::exec(env, v),
-        Command::Stop(v) => stop::exec(env, v),
-        Command::Toolchain(v) => toolchain::exec(env, v),
-        Command::Upgrade(v) => upgrade::exec(env, v),
-        Command::Wallet(v) => wallet::exec(env, v),
+        DfxCommand::Beta(v) => beta::exec(env, v),
+        DfxCommand::Bootstrap(v) => bootstrap::exec(env, v),
+        DfxCommand::Build(v) => build::exec(env, v),
+        DfxCommand::Cache(v) => cache::exec(env, v),
+        DfxCommand::Canister(v) => canister::exec(env, v),
+        DfxCommand::Deploy(v) => deploy::exec(env, v),
+        DfxCommand::Deps(v) => deps::exec(env, v),
+        DfxCommand::Diagnose(v) => diagnose::exec(env, v),
+        DfxCommand::Fix(v) => fix::exec(env, v),
+        DfxCommand::Extension(v) => extension::exec(env, v),
+        // DfxCommand::ExtensionRun(v) => extension::run::exec(env, v.into()),
+        DfxCommand::Generate(v) => generate::exec(env, v),
+        DfxCommand::Identity(v) => identity::exec(env, v),
+        DfxCommand::Info(v) => info::exec(env, v),
+        DfxCommand::LanguageServices(v) => language_service::exec(env, v),
+        DfxCommand::Ledger(v) => ledger::exec(env, v),
+        DfxCommand::New(v) => new::exec(env, v),
+        DfxCommand::Nns(v) => nns::exec(env, v),
+        DfxCommand::Ping(v) => ping::exec(env, v),
+        DfxCommand::Quickstart(v) => quickstart::exec(env, v),
+        DfxCommand::Remote(v) => remote::exec(env, v),
+        DfxCommand::Replica(v) => replica::exec(env, v),
+        DfxCommand::Schema(v) => schema::exec(v),
+        DfxCommand::Sns(v) => sns::exec(env, v),
+        DfxCommand::Start(v) => start::exec(env, v),
+        DfxCommand::Stop(v) => stop::exec(env, v),
+        DfxCommand::Toolchain(v) => toolchain::exec(env, v),
+        DfxCommand::Upgrade(v) => upgrade::exec(env, v),
+        DfxCommand::Wallet(v) => wallet::exec(env, v),
     }
 }
 
-pub fn exec_without_env(cmd: Command) -> DfxResult {
+pub fn exec_without_env(cmd: DfxCommand) -> DfxResult {
     match cmd {
-        Command::Schema(v) => schema::exec(v),
+        DfxCommand::Schema(v) => schema::exec(v),
         _ => bail!("Cannot execute this command without environment."),
     }
 }
+
+impl Subcommand for DfxCommand {
+    fn augment_subcommands(cmd: clap::Command) -> clap::Command {
+        let extension_subcmds = ExtensionManager::new(dfx_version(), false).map_or(vec![], |mgr| {
+            mgr.list_installed_extensions()
+                .unwrap_or_default()
+                .into_iter()
+                .map(|ext| ext.into_clap_command(&mgr))
+                .collect::<Vec<Command>>()
+        });
+
+        let mut cmd = cmd
```

Comment:
Does this help?
```
fn DfxCommand_with_extensions() -> DfxResult<Command> {

    let cli = Command::new("dfx with extensions")
        .subcommand(
            Command::new("nns").about("about nns")
        )
        .subcommand(
            Command::new("sns").about("about sns"));
    let cli = DfxCommand::augment_subcommands(cli);
    Ok(cli)
}
```

Reply:
I did something similar but not this exactly.

I tried this approach but couldn't make it work because I couldn't plug the `Command` back into `CliOpts` (because it results in an error: `duplicate command "beta"` (it duplicates all subcommand)). The example in the link parses from an enum, whereas dfx parses from a struct that contains an enum, therefore we cannot just call `parse` on enum, because we won't be able to get top level args (verbose,version,etc).